### PR TITLE
sql: support more formats for date component of timestamp

### LINKED ISSUE #3109: Extend support text -> date casts for dates in the YYYYMMDD format
Add support for text -> date casts for dates in the `YYYYMMDD` format. That's something that PostgreSQL supports, so we should too. It's much simpler than adding support for the full `to_date` function tracked in #3080, and unblocks a user who currently has dates in that format.

### PR DESCRIPTION
**Pull request by sploiselle**
Closes #3109

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/3128)

### CODE REVIEW

File: src/repr/scalar/strconv.rs [hunk 1]
```diff
@@ -200,6 +200,14 @@ fn parse_timestamp_string(s: &str) -> Result<(NaiveDate, NaiveTime, i64), failur
 
     let (ts_string, tz_string) = crate::datetime::split_timestamp_string(s);
 
+    // Reformat date strings into desired form of 'YYYY-MM-DD', including:
+    // - YYYY MM-DD
+    // - YYYY MM DD
+    // - YYYYMMDD
+    let date_formatter =
+        regex::Regex::new(r"^(?P<y>\d{4})\s*?(?P<m>\d{2})(?:\-|\s+)?(?P<d>\d{2})").unwrap();
```
Comment (quodlibetor) [inline src/repr/scalar/strconv.rs:208]: regex's [should be stuck in a lazy static](https://docs.rs/regex/1.3.7/regex/#example-avoid-compiling-the-same-regex-in-a-loop), even inside a function. They are very expensive to build, and unlike other languages they aren't cached in rust:

```suggestion
lazy_static! {
    static ref DATE_FORMATTER: Regex =
        regex::Regex::new(r"^(?P<y>\d{4})\s*?(?P<m>\d{2})(?:\-|\s+)?(?P<d>\d{2})").unwrap();
}
```

**Comment by benesch** [comment]
This doesn't need to block this PR, but there is something architecturally wrong here if this is the solution? Seems like something that the date/time tokenizer should handle.

**Comment by sploiselle** [comment]
@benesch The date-time parser is hardcoded to expect the one format we support, i.e. `YYYY-MM-DD`; I can refactor that to be more flexible, but will land this right now just to unblock the person waiting on this support.

**Comment by benesch** [comment]
Yep, sounds good!

**Pull request closed without merging**